### ISSUE TITLE
Allow PHPStan to understand conditional return types with generics

### ISSUE THREAD

**Issue by InvisibleSmiley** [laminas/laminas-servicemanager#153]
### Feature Request

|    Q        |   A
|------------ | ------
| New Feature | no
| RFC         | no
| BC Break    | no

#### Summary

PHPStan does not understand some conditional return types which involve generics. As a result, the inferred return type is "object" rather than the class of the input parameter class-string.

Corresponding PHPStan issue: https://github.com/phpstan/phpstan/issues/8049

Solution template: use `class-string<T>` instead of just class-string in @psalm-return.

Occurrences in code:
- ServiceLocatorInterface::build
- AbstractPluginManager::get

**Comment by Ocramius** [comment]
> Solution template: use class-string instead of just class-string in @psalm-return.

Could you rephrase, please?

**Comment by InvisibleSmiley** [comment]
> > Solution template: use class-string instead of just class-string in @psalm-return.
> 
> Could you rephrase, please?

Looks like you cannot simply put something in angle brackets here ;) Edited summary accordingly.

**Comment by Ocramius** [comment]
Aha... could you suggest a patch perhaps? Me and @gsteel are usually quick at picking these up, if valid :)

**Comment by InvisibleSmiley** [comment]
Opened a PR against 3.12.x where this was first introduced.
Not sure how to link the PR to this issue though.

**Comment by Ocramius** [comment]
If it's an improvement, we'll target a later branch: not a problem to retarget though, we can do it, after review  :)
